### PR TITLE
Temporarily accept [i|u][32|size] suffixes on a tuple index and warn

### LINKED ISSUE #59553: Tuple indexing regression
https://github.com/rust-lang/rust/pull/59421 breaks some code generated by proc macros because [quote converts integer to a tokenstream with a suffix](https://github.com/dtolnay/quote/blob/0.6.11/src/to_tokens.rs#L114-L145).

Proc macro:

```rust
extern crate proc_macro;
use proc_macro::TokenStream;
use quote::quote;

#[proc_macro]
pub fn foo(_: TokenStream) -> TokenStream {
    let index = 0;
    TokenStream::from(quote! {
        struct Foo(());
        fn bar() {
            let x = Foo(());
            x.#index
        }
    })
}
```

Calling code:

```rust
foo!();
```

Error:

```text
error: suffixes on a tuple index are invalid
 --> src\lib.rs:3:1
  |
3 | foo!();
  | ^^^^^^^ invalid suffix `i32`
```

It can be avoided by using `syn::Index::from` or `proc_macro::Literal::*_unsuffixed`.

I have already fixed my library (https://github.com/taiki-e/pin-project/commit/5f5cc3508a5eb842d8cb324f55b652fc89070e9b), but it can happen elsewhere.

### LINKED ISSUE #60138: suffixes on a tuple index are invalid
from a crater run: 
```
error: suffixes on a tuple index are invalid
[INFO] [stderr]    --> tests/common/mod.rs:133:10
[INFO] [stderr]     |
[INFO] [stderr] 133 | #[derive(Interact)]
[INFO] [stderr]     |          ^^^^^^^^ invalid suffix `usize`
[INFO] [stderr] 
[INFO] [stderr] error: proc-macro derive produced unparseable tokens
[INFO] [stderr]    --> tests/common/mod.rs:133:10
[INFO] [stderr]     |
[INFO] [stderr] 133 | #[derive(Interact)]
[INFO] [stderr]     |          ^^^^^^^^
[INFO] [stderr] 
```

crate affected: interact (0.3.3), named-tuple (0.1.2), photonix (0.1.1), serde_tuple (0.2.2), susy-codec (3.1.0), thrift (0.12.0), zeroize (0.6.0)
authors: @da-x, @flier, @bsoptei, @kardeiz,  @SUSYLABS, @apache, @tony-iqlusion
logs:
https://crater-reports.s3.amazonaws.com/beta-1.35-1/beta-2019-04-11/reg/interact-0.3.3/log.txt
https://crater-reports.s3.amazonaws.com/beta-1.35-1/beta-2019-04-11/reg/named_tuple-0.1.2/log.txt
https://crater-reports.s3.amazonaws.com/beta-1.35-1/beta-2019-04-11/reg/photonix-0.1.1/log.txt
https://crater-reports.s3.amazonaws.com/beta-1.35-1/beta-2019-04-11/reg/serde_tuple-0.2.2/log.txt
https://crater-reports.s3.amazonaws.com/beta-1.35-1/beta-2019-04-11/reg/susy-codec-3.1.0/log.txt
https://crater-reports.s3.amazonaws.com/beta-1.35-1/beta-2019-04-11/reg/thrift-0.12.0/log.txt,
https://crater-reports.s3.amazonaws.com/beta-1.35-1/beta-2019-04-11/reg/zeroize-0.6.0/log.txt

probable duplicate of #59418

### PR DESCRIPTION
**Pull request by estebank**
Fix #60138.

#59553 will need to be kept open to track the change back to rejecting this code a few versions down thee line.

### CODE REVIEW

**Comment by estebank** [comment]
I tried checking for _only_ spans coming through the macro system, but because the error is pointing at the span for the proc macro itself, it is a regular span and we have no way to differentiate from regular field access.

The warning is unsilenceable _on purpose_.

**Comment by estebank** [comment]
CC @varkor @eddyb @Centril @nikomatsakis

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -1145,9 +1145,34 @@ impl<'a> Parser<'a> {
...
+                {
+                    // #59553: warn instead of reject out of hand to allow the fix to percolate
+                    // through the ecosystem when people fix their macros
+                    let mut err = self.struct_span_warn(
+                        sp,
+                        &format!("suffixes on {} are invalid", kind),
+                    );
+                    err.note(&format!(
+                        "`{}` is *temporarily* accepted on tuple index fields as it was \
+                         incorrectly accepted on stable for a few releases",
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: This affects my opinion, by the way -- what is the range of time in which it was accepted? I was assuming it has been accepted forever.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -1145,9 +1145,34 @@ impl<'a> Parser<'a> {
...
+                        text,
+                    ));
+                    err.help(
+                        "on proc macros, you'll want to use `syn::Index::from` or \
+                         `proc_macro::Literal::*_unsuffixed` for code that will desugar \
+                         to tuple field access",
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: 👍 to giving some explicit help here!

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -1145,9 +1145,34 @@ impl<'a> Parser<'a> {
...
+                    );
+                    err.note(
+                        "for more context, see https://github.com/rust-lang/rust/issues/59553",
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: Can we redirect to https://github.com/rust-lang/rust/issues/60210, which I just created. It's a "clean" issue that describes just this change.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
r=me if we link to the new issue I just filed

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (Centril) [reply src/libsyntax/parse/parser.rs]: Bisected (using godbolt) to 1.27 being the point at which it was allowed;
Test program:
```rust
fn main() {
    let x = (0,);
    x.0usize;
}
```

**Pull request merged**